From 7672009493136596f07e97d0308cba1afba987b8 Mon Sep 17 00:00:00 2001
From: Ethan Cotterell <ethancotterell@gmail.com>
Date: Thu, 20 May 2021 23:31:40 +0100
Subject: [PATCH] patches: icmp_echo ip headers patch

---
 libicmp/icmp_echo.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/libicmp/icmp_echo.c b/libicmp/icmp_echo.c
index f363c28d..0b9a42b7 100644
--- a/libicmp/icmp_echo.c
+++ b/libicmp/icmp_echo.c
@@ -26,7 +26,7 @@
 
 #include <netinet/in_systm.h>
 #include <netinet/in.h>
-#include <netinet/ip.h>
+#include <lwip/ip.h>
 /*#include <netinet/ip_icmp.h> -- deliberately not including this */
 #include <arpa/inet.h>
 #include <icmp.h>
@@ -52,16 +52,16 @@ icmp_generic_encode (unsigned char * buffer, size_t bufsize, int type, int ident
 
 int
 icmp_generic_decode (unsigned char * buffer, size_t bufsize,
-		     struct ip **ipp, icmphdr_t ** icmpp)
+		     struct ip_hdr **ipp, icmphdr_t ** icmpp)
 {
   size_t hlen;
   unsigned short cksum;
-  struct ip *ip;
+  struct ip_hdr *ip;
   icmphdr_t *icmp;
 
   /* IP header */
-  ip = (struct ip *) buffer;
-  hlen = ip->ip_hl << 2;
+  ip = (struct ip_hdr *) buffer;
+  hlen = IPH_HL(ip);
   if (bufsize < hlen + ICMP_MINLEN)
     return -1;
 
@@ -89,7 +89,7 @@ icmp_echo_encode (unsigned char * buffer, size_t bufsize, int ident, int seqno)
 
 int
 icmp_echo_decode (unsigned char * buffer, size_t bufsize,
-		  struct ip **ipp, icmphdr_t ** icmpp)
+		  struct ip_hdr **ipp, icmphdr_t ** icmpp)
 {
   return icmp_generic_decode (buffer, bufsize, ipp, icmpp);
 }
-- 
2.25.1

