From 1ad9f08917ad54d20c3c67d4d8abf8d2b8f4c51d Mon Sep 17 00:00:00 2001
From: Ethan Cotterell <ethancotterell@gmail.com>
Date: Thu, 20 May 2021 23:10:53 +0100
Subject: [PATCH] patches: ping_address ip struct

---
 ping/ping_address.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/ping/ping_address.c b/ping/ping_address.c
index 7163ddd6..0ef58365 100644
--- a/ping/ping_address.c
+++ b/ping/ping_address.c
@@ -48,10 +48,10 @@

 static int recv_address (int code, void *closure,
 			 struct sockaddr_in *dest, struct sockaddr_in *from,
-			 struct ip *ip, icmphdr_t * icmp, int datalen);
+			 struct ip_hdr *ip, icmphdr_t * icmp, int datalen);
 static void print_address (int dupflag, void *closure,
 			   struct sockaddr_in *dest, struct sockaddr_in *from,
-			   struct ip *ip, icmphdr_t * icmp, int datalen);
+			   struct ip_hdr *ip, icmphdr_t * icmp, int datalen);
 static int address_finish (void);

 int
@@ -62,8 +62,10 @@ ping_address (char *hostname)
   ping_set_packetsize (ping, ICMP_MASKLEN);
   ping_set_count (ping, 1);

-  if (ping_set_dest (ping, hostname))
-    error (EXIT_FAILURE, 0, "unknown host");
+  if (ping_set_dest (ping, hostname)) {
+		fprintf(stderr, "unknown host (%s) \n", hostname);
+		exit(1);
+	}

   printf ("PING %s (%s): sending address mask request\n",
 	  ping->ping_hostname, inet_ntoa (ping->ping_dest.ping_sockaddr.sin_addr));
@@ -75,7 +77,7 @@ ping_address (char *hostname)
 int
 recv_address (int code, void *closure,
 	      struct sockaddr_in *dest, struct sockaddr_in *from,
-	      struct ip *ip, icmphdr_t * icmp, int datalen)
+	      struct ip_hdr *ip, icmphdr_t * icmp, int datalen)
 {
   switch (code)
     {
@@ -94,7 +96,7 @@ void
 print_address (int dupflag, void *closure _GL_UNUSED_PARAMETER,
 	       struct sockaddr_in *dest _GL_UNUSED_PARAMETER,
 	       struct sockaddr_in *from,
-	       struct ip *ip _GL_UNUSED_PARAMETER,
+	       struct ip_hdr *ip _GL_UNUSED_PARAMETER,
 	       icmphdr_t * icmp, int datalen)
 {
   struct in_addr addr;
--
2.25.1
