From b639965738bbe5a61f187312dac1e176c3c4711e Mon Sep 17 00:00:00 2001
From: Ethan Cotterell <ethancotterell@gmail.com>
Date: Mon, 7 Jun 2021 18:57:02 +0100
Subject: [PATCH 04/15] ping: fix ip_hdr struct

---
 ping/ping_impl.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ping/ping_impl.h b/ping/ping_impl.h
index 5c9f5cca..a229e00b 100644
--- a/ping/ping_impl.h
+++ b/ping/ping_impl.h
@@ -31,4 +31,4 @@ extern size_t data_length;
 extern int ping_run (PING * ping, int (*finish) ());
 extern int ping_finish (void);
 extern void print_icmp_header (struct sockaddr_in *from,
-			       struct ip *ip, icmphdr_t * icmp, int len);
+			       struct ip_hdr *ip, icmphdr_t * icmp, int len);
-- 
2.25.1

