From bbbd55edff2cf7b56d4422d2c08a813f6560dc3d Mon Sep 17 00:00:00 2001
From: Ethan Cotterell <ethancotterell@gmail.com>
Date: Mon, 7 Jun 2021 19:02:29 +0100
Subject: [PATCH 05/15] ping: fix libping ip_hdr struct

---
 ping/libping.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/ping/libping.c b/ping/libping.c
index 5ee33b9f..9ce69fc2 100644
--- a/ping/libping.c
+++ b/ping/libping.c
@@ -190,11 +190,11 @@ ping_xmit (PING * p)
 static int
 my_echo_reply (PING * p, icmphdr_t * icmp)
 {
-  struct ip *orig_ip = &icmp->icmp_ip;
+  struct ip_hdr *orig_ip = &icmp->icmp_ip;
   icmphdr_t *orig_icmp = (icmphdr_t *) (orig_ip + 1);
 
-  return (orig_ip->ip_dst.s_addr == p->ping_dest.ping_sockaddr.sin_addr.s_addr
-	  && orig_ip->ip_p == IPPROTO_ICMP
+  return (orig_ip->dest.addr == p->ping_dest.ping_sockaddr.sin_addr.s_addr
+	  && orig_ip->_proto == IPPROTO_ICMP
 	  && orig_icmp->icmp_type == ICMP_ECHO
 	  && (ntohs (orig_icmp->icmp_id) == p->ping_ident || useless_ident));
 }
@@ -205,7 +205,7 @@ ping_recv (PING * p)
   socklen_t fromlen = sizeof (p->ping_from.ping_sockaddr);
   int n, rc;
   icmphdr_t *icmp;
-  struct ip *ip;
+  struct ip_hdr *ip;
   int dupflag;
 
   n = recvfrom (p->ping_fd,
-- 
2.25.1

