From bc63622a9ff0a7336eb6f62e0145d6c3ebc90055 Mon Sep 17 00:00:00 2001
From: Ethan Cotterell <ethancotterell@gmail.com>
Date: Fri, 21 May 2021 08:43:29 +0100
Subject: [PATCH] patches: ping_timestamp ip headers patch

---
 ping/ping_timestamp.c | 18 ++++++++++--------
 1 file changed, 10 insertions(+), 8 deletions(-)

diff --git a/ping/ping_timestamp.c b/ping/ping_timestamp.c
index b32c36f1..0590b77d 100644
--- a/ping/ping_timestamp.c
+++ b/ping/ping_timestamp.c
@@ -47,10 +47,10 @@
 
 static int recv_timestamp (int code, void *closure,
 			   struct sockaddr_in *dest, struct sockaddr_in *from,
-			   struct ip *ip, icmphdr_t * icmp, int datalen);
+			   struct ip_hdr *ip, icmphdr_t * icmp, int datalen);
 static void print_timestamp (int dupflag, void *closure,
 			     struct sockaddr_in *dest,
-			     struct sockaddr_in *from, struct ip *ip,
+			     struct sockaddr_in *from, struct ip_hdr *ip,
 			     icmphdr_t * icmp, int datalen);
 static int timestamp_finish (void);
 
@@ -61,8 +61,10 @@ ping_timestamp (char *hostname)
   ping_set_event_handler (ping, recv_timestamp, NULL);
   ping_set_packetsize (ping, ICMP_TSLEN);
 
-  if (ping_set_dest (ping, hostname))
-    error (EXIT_FAILURE, 0, "unknown host");
+  if (ping_set_dest (ping, hostname)) {
+		fprintf(stderr, "unknown host '%s'\n", hostname);
+		exit(1);
+	}
 
   printf ("PING %s (%s): sending timestamp requests\n",
 	  ping->ping_hostname, inet_ntoa (ping->ping_dest.ping_sockaddr.sin_addr));
@@ -74,7 +76,7 @@ ping_timestamp (char *hostname)
 int
 recv_timestamp (int code, void *closure,
 		struct sockaddr_in *dest, struct sockaddr_in *from,
-		struct ip *ip, icmphdr_t * icmp, int datalen)
+		struct ip_hdr *ip, icmphdr_t * icmp, int datalen)
 {
   switch (code)
     {
@@ -91,10 +93,10 @@ recv_timestamp (int code, void *closure,
 
 
 void
-print_timestamp (int dupflag, void *closure _GL_UNUSED_PARAMETER,
-		 struct sockaddr_in *dest _GL_UNUSED_PARAMETER,
+print_timestamp (int dupflag, void *closure,
+		 struct sockaddr_in *dest,
 		 struct sockaddr_in *from,
-		 struct ip *ip _GL_UNUSED_PARAMETER,
+		 struct ip_hdr *ip,
 		 icmphdr_t * icmp, int datalen)
 {
   char timestr[16];
-- 
2.25.1

