From fdc65a9e82069e4e2c2c3ac4ce684c495eb21c4f Mon Sep 17 00:00:00 2001
From: Ethan Cotterell <ethancotterell@gmail.com>
Date: Mon, 7 Jun 2021 21:58:30 +0100
Subject: [PATCH 08/13] ping: fix ip struct ping_timestamp

---
 ping/ping_timestamp.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/ping/ping_timestamp.c b/ping/ping_timestamp.c
index b32c36f1..ac03919b 100644
--- a/ping/ping_timestamp.c
+++ b/ping/ping_timestamp.c
@@ -47,10 +47,10 @@
 
 static int recv_timestamp (int code, void *closure,
 			   struct sockaddr_in *dest, struct sockaddr_in *from,
-			   struct ip *ip, icmphdr_t * icmp, int datalen);
+			   struct ip_hdr *ip, icmphdr_t * icmp, int datalen);
 static void print_timestamp (int dupflag, void *closure,
 			     struct sockaddr_in *dest,
-			     struct sockaddr_in *from, struct ip *ip,
+			     struct sockaddr_in *from, struct ip_hdr *ip,
 			     icmphdr_t * icmp, int datalen);
 static int timestamp_finish (void);
 
@@ -61,8 +61,10 @@ ping_timestamp (char *hostname)
   ping_set_event_handler (ping, recv_timestamp, NULL);
   ping_set_packetsize (ping, ICMP_TSLEN);
 
-  if (ping_set_dest (ping, hostname))
-    error (EXIT_FAILURE, 0, "unknown host");
+  if (ping_set_dest (ping, hostname)) {
+		fprintf(stderr, "unknown host\n");
+		exit(EXIT_FAILURE);
+	}
 
   printf ("PING %s (%s): sending timestamp requests\n",
 	  ping->ping_hostname, inet_ntoa (ping->ping_dest.ping_sockaddr.sin_addr));
@@ -74,7 +76,7 @@ ping_timestamp (char *hostname)
 int
 recv_timestamp (int code, void *closure,
 		struct sockaddr_in *dest, struct sockaddr_in *from,
-		struct ip *ip, icmphdr_t * icmp, int datalen)
+		struct ip_hdr *ip, icmphdr_t * icmp, int datalen)
 {
   switch (code)
     {
@@ -94,7 +96,7 @@ void
 print_timestamp (int dupflag, void *closure _GL_UNUSED_PARAMETER,
 		 struct sockaddr_in *dest _GL_UNUSED_PARAMETER,
 		 struct sockaddr_in *from,
-		 struct ip *ip _GL_UNUSED_PARAMETER,
+		 struct ip_hdr *ip _GL_UNUSED_PARAMETER,
 		 icmphdr_t * icmp, int datalen)
 {
   char timestr[16];
-- 
2.25.1

