From 80ca8061f7684b8ef91dc1911f2a0c72a7bc7690 Mon Sep 17 00:00:00 2001
From: Ethan Cotterell <ethancotterell@gmail.com>
Date: Fri, 21 May 2021 13:28:37 +0100
Subject: [PATCH 11/14] patches: cmdtab

---
 ftp/cmdtab.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ftp/cmdtab.c b/ftp/cmdtab.c
index fe1cf397..e9b0f7a3 100644
--- a/ftp/cmdtab.c
+++ b/ftp/cmdtab.c
@@ -53,7 +53,7 @@
 #include <stdio.h>
 #include <string.h>
 
-#include "ftp_var.h"
+#include <ftp_var.h>
 
 /*
  * User FTP -- Command Tables.
-- 
2.25.1

