From 4c4c1420a256c8bc408f1099c8613619a16b304b Mon Sep 17 00:00:00 2001
From: Ethan Cotterell <ethancotterell@gmail.com>
Date: Fri, 21 May 2021 13:34:23 +0100
Subject: [PATCH] patches: ftp domacro patches

---
 ftp/domacro.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ftp/domacro.c b/ftp/domacro.c
index 931e9ee7..f48b6594 100644
--- a/ftp/domacro.c
+++ b/ftp/domacro.c
@@ -55,7 +55,7 @@
 #include <stdlib.h>
 #include <string.h>
 
-#include "ftp_var.h"
+#include <ftp_var.h>
 
 /* Increase allocated length of `*startl' with `add'
  * beyond `*track',  should present length not suffice.
-- 
2.25.1

